src/examples/source-3-non-det/liars.js

```diff
@@ -0,0 +1,31 @@
+/*
+  SICP JS Exercise 4.33
+
+  This file contains a solution to the Liars' puzzle
+*/
+
+function distinct(items) {
+  return is_null(items)
+      ? true
+      : is_null(tail(items))
+      ? true
+      : is_null(member(head(items), tail(items)))
+          ? distinct(tail(items))
+          : false;
+}
+
+const kitty = amb(1, 2, 3, 4, 5);
+const mary = amb(1, 2, 3, 4, 5);
+require(amb(kitty === 2, mary === 4));
```

Comment:
To reduce the search space, I have done the `require` checks immediately once the information needed for the check is available.

Reply:
Great.